src/test/groovy/com/streamr/client/SubscriptionSpec.groovy

```diff
@@ -0,0 +1,226 @@
+                assert msg.toJson() == message.toJson()
+                done = true
+            }
+        })
+        sub.handleMessage(msg, 'session-token')
+        then:
+        done
+    }
+
+    void "calls the handler once for each message in order"() {
+        ArrayList<StreamMessage> msgs = new ArrayList<>()
+        for (int i=0;i<5;i++) {
+            msgs.add(createMessage((long)i, 0, null, 0))
+        }
+        ArrayList<StreamMessage> received = new ArrayList<>()
+        int counter = 0
+        when:
+        Subscription sub = new Subscription(msg.getStreamId(), msg.getStreamPartition(), new MessageHandler() {
+            @Override
+            void onMessage(Subscription sub, StreamMessage message) {
+                assert msgs.get(counter).toJson() == message.toJson()
+                received.add(message)
+                counter++
+                if (counter == 5) {
+                    done = true
+                }
+            }
+        })
+        for (int i=0;i<5;i++) {
+            sub.handleMessage(msgs.get(i), 'session-token')
+        }
+        then:
+        done
+    }
+
+    void "does not handle messages (queued) during resending"() {
+        when:
+        Subscription sub = new Subscription(msg.getStreamId(), msg.getStreamPartition(), new MessageHandler() {
+            @Override
+            void onMessage(Subscription sub, StreamMessage message) {
+                throw new Exception("Shouldn't handle this message!")
+            }
+        })
+        sub.setResending(true)
+        sub.handleMessage(msg, 'session-token')
+        then:
+        assert true
+    }
+
+    void "handles messages during resending if isResend=true"() {
+        when:
+        Subscription sub = new Subscription(msg.getStreamId(), msg.getStreamPartition(), new MessageHandler() {
+            @Override
+            void onMessage(Subscription sub, StreamMessage message) {
+                assert msg.toJson() == message.toJson()
+                done = true
+            }
+        })
+        sub.setResending(true)
+        sub.handleMessage(msg, 'session-token', true)
+        then:
+        done
+    }
+
+    void "ignores duplicate messages"() {
+        int counter = 0
+        when:
+        Subscription sub = new Subscription(msg.getStreamId(), msg.getStreamPartition(), new MessageHandler() {
+            @Override
+            void onMessage(Subscription sub, StreamMessage message) {
+                assert msg.toJson() == message.toJson()
+                counter++
+                done = true
+                if (counter == 2) {
+                    throw new Exception("Shouldn't handle this duplicate message!")
+                }
+            }
+        })
+        sub.handleMessage(msg, 'session-token')
+        sub.handleMessage(msg, 'session-token')
+        then:
+        done
+    }
+
+    void "returns a ResendRangeRequest if a gap is detected"() {
```

Comment:
When moving to declared exception model, you should get nice related when-then blocks here:
```
// setup
// ...
when:
sub.handleMessage(msg1)
then:
noExceptionThrown()

when:
sub.handleMessage(msg4)
then:
GapDetectedException ex = thrown(GapDetectedException)
ex.from == ... // check exception fields
```